### PR TITLE
RGW\bucket notification tests: end-point agnostic + refactoring

### PR DESCRIPTION
**Pull request by AliMasarweh**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by yuvalif** [comment]
@AliMasarweh please open a tracker for this PR, and mark as "squid" backport.
* it is a great improvement to our notification tests
* without it, it wil be difficult to backport any other test fixes to squid

File: src/test/rgw/bucket_notification/test_bn.py [hunk 1]
```diff
@@ -1232,9 +1237,13 @@ def notification_push(endpoint_type, conn, account=None):
         host = get_ip_http()
         port = random.randint(10000, 20000)
         # start an http server in a separate thread
-        receiver = HTTPServerWithEvents((host, port))
+        receiver = HTTPServerWithEvents((host, port), cloudevents=cloudevents)
         endpoint_address = 'http://'+host+':'+str(port)
-        endpoint_args = 'push-endpoint='+endpoint_address
+        endpoint_args = ''
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:1242]: no need to initialize `endpoint_args` there is not scope in pythoon

File: src/test/rgw/bucket_notification/test_bn.py [hunk 2]
```diff
@@ -1291,6 +1300,8 @@ def notification_push(endpoint_type, conn, account=None):
 
     # create objects in the bucket
     number_of_objects = 100
+    if cloudevents:
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:1303]: why is this different?

File: src/test/rgw/bucket_notification/test_bn.py [hunk 3]
```diff
@@ -1717,38 +1652,54 @@ def test_ps_s3_opaque_data_on_master():
...
-    conn = connection()
-    zonegroup = get_config_zonegroup()
 
-    # create random port for the http server
-    host = get_ip()
-    port = random.randint(10000, 20000)
-    # start an http server in a separate thread
-    number_of_objects = 10
-    http_server = HTTPServerWithEvents((host, port))
+def lifecycle(endpoint_type, conn, number_of_objects, topic_events, create_thread, rules_creator, record_events,
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:1656]: why do we need the thread creator function as a parameter?

File: src/test/rgw/bucket_notification/test_bn.py [hunk 4]
```diff
@@ -1717,38 +1652,54 @@ def test_ps_s3_opaque_data_on_master():
...
+              expected_abortion=False):
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:1657]: when is this parameter used?

File: src/test/rgw/bucket_notification/test_bn.py [hunk 5]
```diff
@@ -1818,99 +1758,82 @@ def test_ps_s3_lifecycle_on_master():
...
-        assert key_name in event['Records'][0]['s3']['object']['key']
 
-    # cleanup
-    topic_conf.del_config()
-    s3_notification_conf.del_config(notification=notification_name)
-    # delete the bucket
-    conn.delete_bucket(bucket_name)
-    http_server.close()
 
 def ps_s3_creation_triggers_on_master(external_endpoint_address=None, ca_location=None, verify_ssl='true'):
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:1838]: we can probably make that endpoint agnostic as well

Comment (yuvalif) [reply src/test/rgw/bucket_notification/test_bn.py:1838]: i think we can remove the external endpoint logic.
we can say that rabbitmq must be running (similar to how we do that with kafka), and not run it from within the test

Comment (yuvalif) [reply src/test/rgw/bucket_notification/test_bn.py:1838]: and aslo make `test_http_post_object_upload` endpoint agnostic

File: src/test/rgw/bucket_notification/test_bn.py [hunk 6]
```diff
@@ -2266,111 +2204,38 @@ def test_ps_s3_multipart_on_master_http():
 
     # check http receiver
     keys = list(bucket.list())
-    print('total number of objects: ' + str(len(keys)))
-    events = http_server.get_and_reset_events()
-    for event in events:
-        assert_equal(event['Records'][0]['opaqueData'], opaque_data)
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:2272]: * we probably need a separate test for the `opaque_data`
* i think we are allready testing etags elsewhere in a better way

File: src/test/rgw/bucket_notification/test_bn.py [hunk 7]
```diff
@@ -2266,111 +2204,38 @@ def test_ps_s3_multipart_on_master_http():
...
     for key in keys:
         key.delete()
-    [thr.join() for thr in client_threads]
-    topic_conf.del_config()
-    s3_notification_conf.del_config(notification=notification_name)
     # delete the bucket
     conn.delete_bucket(bucket_name)
-    http_server.close()
-
+    receiver.close(task)
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:2217]: you shoudl also check if `task != None` and call `task.stop()`

File: src/test/rgw/bucket_notification/test_bn.py [hunk 8]
```diff
@@ -2266,111 +2204,38 @@ def test_ps_s3_multipart_on_master_http():
...
-    # delete the bucket
-    conn.delete_bucket(bucket_name)
+@attr('amqp_test')
+def test_multipart_ampq():
+    """ test ampq multipart object upload """
+    conn = connection()
+    test_multipart('ampq', conn)
 
 @attr('amqp_test')
 def test_ps_s3_metadata_filter_on_master():
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:2241]: make endpoint agnostic?

Comment (AliMasarweh) [reply src/test/rgw/bucket_notification/test_bn.py:2241]: sure

File: src/test/rgw/bucket_notification/test_bn.py (hunk 2 above)
Comment (AliMasarweh) [reply src/test/rgw/bucket_notification/test_bn.py:1303]: It was 10 before, and cloudevents misses some object if we increase the number

File: src/test/rgw/bucket_notification/test_bn.py (hunk 5 above)
Comment (AliMasarweh) [reply src/test/rgw/bucket_notification/test_bn.py:1838]: does the **ca_location** work for all endpoint types?

File: src/test/rgw/bucket_notification/test_bn.py (hunk 6 above)
Comment (AliMasarweh) [reply src/test/rgw/bucket_notification/test_bn.py:2272]: got it

File: src/test/rgw/bucket_notification/test_bn.py (hunk 7 above)
Comment (AliMasarweh) [reply src/test/rgw/bucket_notification/test_bn.py:2217]: there is not stop function for task

**Comment by yuvalif** [comment]
in a followup PR/commit please split http tests into their own task.
in this directory: https://github.com/ceph/ceph/tree/main/qa/suites/rgw/notifications/tasks
add another directory "http" similar to "others". but with `test_http.yaml` file:
```yaml
tasks:
- notification-tests:
    client.0:
      extra_attr: ["http_test"]
      rgw_server: client.0
```
and in : https://github.com/ceph/ceph/blob/main/qa/tasks/notification_tests.py
add `!http_test` to: https://github.com/ceph/ceph/blob/main/qa/tasks/notification_tests.py#L223C8-L223C37

File: src/test/rgw/bucket_notification/test_bn.py (hunk 5 above)
Comment (yuvalif) [reply src/test/rgw/bucket_notification/test_bn.py:1838]: not really. this test is actually mixing two unrelated things - testing triggers and testing amqp security.
i think you are right here. lets keep that out-of-scope for now

File: src/test/rgw/bucket_notification/test_bn.py (hunk 2 above)
Comment (yuvalif) [reply src/test/rgw/bucket_notification/test_bn.py:1303]: cloudevents are just some extra http headers. if we are missing events there might be an issue.
lets investigate that

File: src/test/rgw/bucket_notification/test_bn.py [hunk 9]
```diff
@@ -823,65 +862,9 @@ def test_ps_s3_topic_with_secret_on_master():
 
 
 @attr('basic_test')
-def test_ps_s3_notification_on_master():
+def test_notification_configuration_on_master():
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:865]: nit:
```suggestion
def test_notification_configuration():
```

**Comment by AliMasarweh** [comment]
passing teuthology run:
https://pulpito.ceph.com/alimasa-2024-05-30_14:26:24-rgw:notifications-wip-alimasa-bn-end-point-distro-default-smithi/

**Comment by AliMasarweh** [comment]
jenkins test api

**Comment by AliMasarweh** [comment]
jenkins test api

**Comment by AliMasarweh** [comment]
jenkins test api

**Comment by AliMasarweh** [comment]
jenkins test windows

**Review by yuvalif — APPROVED** [review]

**Pull request merged**